File: core/src/com/qtech/bubbleblaster/settings/GameSettings.java
package com.qtech.bubbleblaster.settings;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.qtech.bubbleblaster.BubbleBlaster;
import com.qtech.bubbleblaster.common.References;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardOpenOption;
import java.util.Locale;
import java.util.Objects;

public class GameSettings {
    private static final GameSettings INSTANCE = new GameSettings();

    protected final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private JsonObject parsed;

    public GameSettings() {
        reload();
    }

    public synchronized void reload() {
        Gson gson = new Gson();
        File settingsFile = References.SETTINGS_FILE;

        String json = "{}";

        BubbleBlaster.getLogger().info("Reloading settings file...");

        if (!settingsFile.exists()) {
            try {
                BubbleBlaster.getLogger().debug("Write settings file from resource...");
                Files.copy(Objects.requireNonNull(getClass().getResourceAsStream("/settings.json")), settingsFile.toPath());
            } catch (IOException e) {
                BubbleBlaster.getLogger().error("Cannot write settings file from resource: " + settingsFile);
                e.printStackTrace();
            }
        }

        try {
            BubbleBlaster.getLogger().debug("Reading settings file...");
            json = Files.readString(settingsFile.toPath());
        } catch (IOException e) {
            BubbleBlaster.getLogger().error("Cannot read settings file: " + settingsFile.toPath());
            e.printStackTrace();
        }

        boolean changed = false;
        try {
            parsed = gson.fromJson(json, JsonObject.class);
        } catch (Exception e) {
            BubbleBlaster.getLogger().fatal("Cannot read Game-settings file. Traceback follows:");
            e.printStackTrace();
            BubbleBlaster.getInstance().shutdown();
            return;
        }
        if (!parsed.has("lang")) {
            parsed.add("lang", new JsonPrimitive("en_us"));
            changed = true;
        }
        if (!parsed.has("max-bubbles")) {
//            parsed.add("max-bubbles", new JsonPrimitive(Runtime.getRuntime().availableProcessors() * 62));
            parsed.add("max-bubbles", new JsonPrimitive(500));
            changed = true;
        }
        if (!parsed.has("graphics")) {
//            parsed.add("max-bubbles", new JsonPrimitive(Runtime.getRuntime().availableProcessors() * 62));
            JsonObject graphics = new JsonObject();
            graphics.add("enable-antialias", new JsonPrimitive(true));
            graphics.add("enable-text-antialias", new JsonPrimitive(true));

            parsed.add("graphics", graphics);
            changed = true;
        } else {
            JsonObject graphics = parsed.getAsJsonObject("graphics");

            if (!graphics.has("enable-antialias")) {
                graphics.add("enable-antialias", new JsonPrimitive(true));
                changed = true;
            }
            if (!graphics.has("enable-text-antialias")) {
                graphics.add("enable-text-antialias", new JsonPrimitive(true));
                changed = true;
            }

            parsed.add("graphics", graphics);
        }

        if (changed) {
            save();
        }

        BubbleBlaster.getLogger().info("Settings file reloaded!");
    }

    public synchronized void save() {
        File settingsFile = References.SETTINGS_FILE;

        BubbleBlaster.getLogger().info("Saving settings file...");

        String json = gson.toJson(parsed, JsonObject.class);
        try {
            BubbleBlaster.getLogger().info("Writing settings file...");
            Files.writeString(settingsFile.toPath(), json, StandardCharsets.UTF_8, StandardOpenOption.CREATE);
        } catch (IOException e) {
            BubbleBlaster.getLogger().error("Cannot write settings file to: " + settingsFile);
            e.printStackTrace();
        }

        BubbleBlaster.getLogger().info("Settings file saved!");
    }

    public static GameSettings instance() {
        return INSTANCE;
    }

    public int getMaxBubbles() {
        return parsed.getAsJsonPrimitive("max-bubbles").getAsInt();
    }

    public void setMaxBubbles(int value) {
        parsed.add("max-bubbles", new JsonPrimitive(value));
    }

    public String getLanguage() {
        return parsed.getAsJsonPrimitive("lang").getAsString();
    }

    public Locale getLanguageLocale() {
        return Locale.forLanguageTag(getLanguage());
    }

    public void setLanguage(String lang) {
        parsed.add("lang", new JsonPrimitive(Locale.forLanguageTag(lang).getLanguage().toLowerCase()));
    }

    public void setLanguage(Locale locale) {
        parsed.add("lang", new JsonPrimitive(locale.getLanguage().toLowerCase()));
    }

    public boolean isAntialiasEnabled() {
        return parsed.getAsJsonObject("graphics").getAsJsonPrimitive("enable-antialias").getAsBoolean();
    }

    public boolean isTextAntialiasEnabled() {
        return parsed.getAsJsonObject("graphics").getAsJsonPrimitive("enable-text-antialias").getAsBoolean();
    }

    public void setAntialiasEnabled(boolean b) {
        JsonObject graphics = parsed.getAsJsonObject("graphics");
        graphics.add("enable-antialias", new JsonPrimitive(b));
        parsed.add("graphics", graphics);
    }

    public void setTextAntialiasEnabled(boolean b) {
        JsonObject graphics = parsed.getAsJsonObject("graphics");
        graphics.add("enabled-text-antialias", new JsonPrimitive(b));
        parsed.add("graphics", graphics);
    }
}

File: src/main/kotlin/qtech/bubbles/settings/GameSettings.kt
@file:Suppress("unused")

package qtech.bubbles.settings

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import qtech.bubbles.BubbleBlaster
import qtech.bubbles.common.References
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import java.util.*

open class GameSettings {
    protected val gson: Gson = GsonBuilder().setPrettyPrinting().create()
    private lateinit var parsed: JsonObject
    @Synchronized
    fun reload() {
        val gson = Gson()
        val settingsFile = References.SETTINGS_FILE
        var json: String? = "{}"
        BubbleBlaster.logger.info("Reloading settings file...")
        if (!settingsFile.exists()) {
            try {
                BubbleBlaster.logger.debug("Write settings file from resource...")
                Files.copy(Objects.requireNonNull(javaClass.getResourceAsStream("/settings.json")), settingsFile.toPath())
            } catch (e: IOException) {
                BubbleBlaster.logger.error("Cannot write settings file from resource: $settingsFile")
                e.printStackTrace()
            }
        }
        try {
            BubbleBlaster.logger.debug("Reading settings file...")
            json = Files.readString(settingsFile.toPath())
        } catch (e: IOException) {
            BubbleBlaster.logger.error("Cannot read settings file: " + settingsFile.toPath())
            e.printStackTrace()
        }
        var changed = false
        parsed = try {
            gson.fromJson(json, JsonObject::class.java)
        } catch (e: Exception) {
            BubbleBlaster.logger.fatal("Cannot read Game-settings file. Traceback follows:")
            e.printStackTrace()
            BubbleBlaster.instance.shutdown()
            return
        }
        if (!parsed.has("lang")) {
            parsed.add("lang", JsonPrimitive("en_us"))
            changed = true
        }
        if (!parsed.has("max-bubbles")) {
//            parsed.add("max-bubbles", new JsonPrimitive(Runtime.getRuntime().availableProcessors() * 62));
            parsed.add("max-bubbles", JsonPrimitive(500))
            changed = true
        }
        if (!parsed.has("graphics")) {
//            parsed.add("max-bubbles", new JsonPrimitive(Runtime.getRuntime().availableProcessors() * 62));
            val graphics = JsonObject()
            graphics.add("enable-antialias", JsonPrimitive(true))
            graphics.add("enable-text-antialias", JsonPrimitive(true))
            parsed.add("graphics", graphics)
            changed = true
        } else {
            val graphics = parsed.getAsJsonObject("graphics")
            if (!graphics.has("enable-antialias")) {
                graphics.add("enable-antialias", JsonPrimitive(true))
                changed = true
            }
            if (!graphics.has("enable-text-antialias")) {
                graphics.add("enable-text-antialias", JsonPrimitive(true))
                changed = true
            }
            parsed.add("graphics", graphics)
        }
        if (changed) {
            save()
        }
        BubbleBlaster.logger.info("Settings file reloaded!")
    }

    @Synchronized
    fun save() {
        val settingsFile = References.SETTINGS_FILE
        BubbleBlaster.logger.info("Saving settings file...")
        val json = gson.toJson(parsed, JsonObject::class.java)
        try {
            BubbleBlaster.logger.info("Writing settings file...")
            Files.writeString(settingsFile.toPath(), json, StandardCharsets.UTF_8, StandardOpenOption.CREATE)
        } catch (e: IOException) {
            BubbleBlaster.logger.error("Cannot write settings file to: $settingsFile")
            e.printStackTrace()
        }
        BubbleBlaster.logger.info("Settings file saved!")
    }

    var maxBubbles: Int
        get() = parsed.getAsJsonPrimitive("max-bubbles").asInt
        set(value) {
            parsed.add("max-bubbles", JsonPrimitive(value))
        }
    var language: String?
        get() = parsed.getAsJsonPrimitive("lang").asString
        set(lang) {
            parsed.add("lang", JsonPrimitive(Locale.forLanguageTag(lang).language.lowercase()))
        }
    val languageLocale: Locale
        get() = Locale.forLanguageTag(language)

    fun setLanguage(locale: Locale) {
        parsed.add("lang", JsonPrimitive(locale.language.lowercase()))
    }

    var isAntialiasEnabled: Boolean
        get() = parsed.getAsJsonObject("graphics").getAsJsonPrimitive("enable-antialias").asBoolean
        set(b) {
            val graphics = parsed.getAsJsonObject("graphics")
            graphics.add("enable-antialias", JsonPrimitive(b))
            parsed.add("graphics", graphics)
        }
    var isTextAntialiasEnabled: Boolean
        get() = parsed.getAsJsonObject("graphics").getAsJsonPrimitive("enable-text-antialias").asBoolean
        set(b) {
            val graphics = parsed.getAsJsonObject("graphics")
            graphics.add("enabled-text-antialias", JsonPrimitive(b))
            parsed.add("graphics", graphics)
        }

    companion object {
        private val INSTANCE = GameSettings()
        @JvmStatic
        fun instance(): GameSettings {
            return INSTANCE
        }
    }

    init {
        reload()
    }
}